src/common/components/dateTimePicker/DateTimePicker.tsx

```diff
@@ -0,0 +1,303 @@
+  message: string
+): string {
+  if (!date || !dateToCompare) {
+    return "";
+  }
+
+  const isError = fnc(date, dateToCompare);
+
+  if (isError) {
+    return message;
+  }
+
+  return "";
+}
+
+function useOnOutsideClick<T extends HTMLElement>(
+  callback: () => void
+): Ref<T | null> {
+  const ref = useRef<null | T>(null);
+
+  useEffect(() => {
+    const listener = (event) => {
+      if (ref.current && !ref.current.contains(event.target)) {
+        callback();
+      }
+    };
+
+    document.addEventListener("mousedown", listener);
+    document.addEventListener("touchstart", listener);
+
+    return () => {
+      document.removeEventListener("mousedown", listener);
+      document.removeEventListener("touchstart", listener);
+    };
+  }, [ref, callback]);
+
+  return ref;
+}
+
+type Props = {
+  id: string;
+  name: string;
+  onChange?: (value: Date) => void;
+  value?: Date;
+  label: string;
+  locale?: Locale;
+  minDate?: Date;
+  maxDate?: Date;
+  minDateErrorMessage?: string;
+  maxDateErrorMessage?: string;
+};
+
+export default function DateTimePicker({
+  id,
+  name,
+  label,
+  onChange,
+  value = null,
+  locale = "fi",
+  minDate,
+  maxDate,
+  minDateErrorMessage,
+  maxDateErrorMessage,
+}: Props) {
+  const dateString = getDate(value);
+  const timeString = getTime(value);
+
+  const { t } = useTranslation("date_time_picker");
+  const [isOpen, setIsOpen] = useState<boolean>(false);
+  const [intermediaryDate, setIntermediaryDate] = useIntermediaryState<
+    string | null
+  >(dateString);
+  const [intermediaryTime, setIntermediaryTime] = useIntermediaryState<
+    string | null
+  >(timeString);
+  const [date, setDate] = useState<Date | null>(value);
+  const buttonRef = useRef(null);
+  const containerRef = useOnOutsideClick<HTMLDivElement>(
+    useCallback(() => setIsOpen(false), [])
+  );
+
+  const handleToggleMenu = () => {
+    setIsOpen((previousValue) => !previousValue);
+  };
+
+  const handleCloseMenuButtonClick = () => {
+    setIsOpen(false);
+  };
+
+  const handleDateChange = (nextDate: string) => {
+    setIntermediaryDate(nextDate);
+    onDatePartChange(nextDate, intermediaryTime);
+  };
+
+  const handleTimeChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const nextTime = event.target.value;
+
+    setIntermediaryTime(nextTime);
+    onDatePartChange(intermediaryDate, nextTime);
+  };
+
+  const onDatePartChange = (date: string, time: string) => {
+    const dateObject = getDateFromDateAndTimeString(date, time);
+    const isValidDate = dateObject ? getIsDateValid(dateObject) : false;
+
+    if (isValidDate) {
+      setDate(dateObject);
+    }
+  };
+
+  const handleOnKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+    switch (e.key) {
+      case "Escape":
+        return setIsOpen(false);
+    }
+  };
+
+  const handleSelectDateClick = () => {
+    onChange(date);
+
+    if (buttonRef.current) {
+      buttonRef.current.focus();
```

Comment:
Not sure which one do you mean would transpile to `buttonRef.current ? buttonRef.current() : null`.

I can't see any reason not to leverage optional chaining in these situations because it makes the code little bit easier to read and `?` can communicate the same thing. Also here we really don't care if `buttonRef.current` is null because we just want to focus it if it is in the DOM.

But again, just my opinion and pretty minor detail

Reply:
I think that optional chaining transpiles to the equivalent of `buttonRef.current ? buttonRef.current() : null` and therefore it's not the same as `onHandler && onHandler()`.

Using `if` is a lot easier to distinct (you can miss the `?`) and it distributes the logic on separates rows. With chaining you check existence and invoke on the same line, with `if` you do not. Personally I don't see either as being clearly better overall, but I can see why you would prefer one or the other in some scenarios.

But I don't think that this is a point where we need to agree or enforce an approach. I don't think it makes a great enough difference to be worth a decision. I'm perfectly fine with seeing chaining and ifs used interchangeably.

Reply:
I think that `?` is not relevant in this situation (and many others), even if you miss it when reading the code it is not a problem and not a important detail.

But yeah, no need to enforce approach here.

And in the first place I commented just to make sure if you have considered using optional chaining and made a conscious decision not to use it.  Also good to have discussion about this, no need to comment about these in the future :D

Reply:
Yeah, sounds good 👍 With that last paragraph I just wanted to make it clear that it's perfectly fine for me if you want to use chaining even if there are places where it's not, as this PR is sort of a coding philosophy discussion now :D